### PR TITLE
2711-catch-rpt-month-year-mismatches

### LINKED ISSUE #2711: Catch rpt_month_year mismatches
**Description:**
As a follow-on to #1613, @ADPennington brought up that a mismatch between the frontend quarter submission and the file contents report_month_year will not trigger a datafile rejection. 

**Acceptance Criteria:** 
_Create a list of functional outcomes that must be achieved to complete this issue_

- [x] DataFile with header RMY mismatch is rejected via pre-check and case_aggregates show as "N/A"
- [x] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 

- Implement `parse_datafile` precheck error similar to section precheck
- Ensure accepted data file for correct rpt_month_year still shows correct case_aggregates

**Notes:**
_Add additional useful information, such as related issues and functionality that isn't covered by this specific issue, and other considerations that will be helpful for anyone reading this_ 

**Supporting Documentation:**
_Please include any relevant log snippets/files/screen shots_

**Open Questions:**
_Please include any questions or decisions that must be made before beginning work or to confidently call this issue complete_

### PR DESCRIPTION
**Pull request by raftmsohani**
## Summary of Changes
Pull request closes #2711 _

Closes the gap between reporting month/year from frontend and month/year of the header in the reporting file.

## How to Test

Start the app by:
```
cd tdrs-backend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d 
cd tdrs-frontend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d
```

1. Open http://localhost:3000/ and sign in.
1. Submit a file. Make sure the reporting year/month/quarter matches the header file.
1. Confirm that the file is submitted successfully.
1. Submit another file and mismatch the frontend reporting year/quarter with the file header.
1. Confirm that the file throws error and is not accepted. 

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] DataFile with header RMY mismatch is rejected via pre-check and case_aggregates show as "N/A"
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [x] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

**Comment by raftmsohani** [comment]
@ADPennington: I have the following two questions:
1. Are we only checking for quarter to match? or do we need to check year as well?
2. Should we account for fiscal quarter when validating the quarter?

**Comment by ADPennington** [comment]
> @ADPennington: I have the following two questions:
> 
> 1. Are we only checking for quarter to match? or do we need to check year as well?

both, and together, i think. this ticket originated [here](https://github.com/raft-tech/TANF-app/pull/2612#discussion_r1334441349). 

> 2. Should we account for fiscal quarter when validating the quarter?

i'm not sure if fiscal quarter can be validated by itself. generally speaking, fiscal year + fiscal quarter selections need to map to the calendar periods in the files. If they dont map correctly, the file should be rejected and not parsed.  i think this looks a bit different by section:

- section 1: FY2024Q1 (frontend selections) means `RPT_MONTH_YEAR` in the T1/M2, T2/M2, and M3/T3 records should have values equal to 202310, 202311 or 202312, and so on... 
- section 2: similar to Section 1 but  for T4/M4 and T5/M5 records
- section 3: FY2024Q1 means `CALENDAR_QUARTER` in the T6/M6 record should have a value equal to 20234, and so on...
- section 4: similar to section 3 but for T7/M7 records

**Comment by raftmsohani** [comment]
> > @ADPennington: I have the following two questions:
> > 
> > 1. Are we only checking for quarter to match? or do we need to check year as well?
> 
> both, and together, i think. this ticket originated [here](https://github.com/raft-tech/TANF-app/pull/2612#discussion_r1334441349).
> 
> > 2. Should we account for fiscal quarter when validating the quarter?
> 
> i'm not sure if fiscal quarter can be validated by itself. generally speaking, fiscal year + fiscal quarter selections need to map to the calendar periods in the files. If they dont map correctly, the file should be rejected and not parsed. i think this looks a bit different by section:
> 
> * section 1: FY2024Q1 (frontend selections) means `RPT_MONTH_YEAR` in the T1/M2, T2/M2, and M3/T3 records should have values equal to 202310, 202311 or 202312, and so on...
> * section 2: similar to Section 1 but  for T4/M4 and T5/M5 records
> * section 3: FY2024Q1 means `CALENDAR_QUARTER` in the T6/M6 record should have a value equal to 20234, and so on...
> * section 4: similar to section 3 but for T7/M7 records

@ADPennington After discussing with the team, we decided to compare the HEADER year and quarter with frontend date, and draft another ticket for further work to compare each field with frontend date.

**Comment by ADPennington** [comment]
> > > @ADPennington: I have the following two questions:
> > > 
> > > 1. Are we only checking for quarter to match? or do we need to check year as well?
> > 
> > 
> > both, and together, i think. this ticket originated [here](https://github.com/raft-tech/TANF-app/pull/2612#discussion_r1334441349).
> > > 2. Should we account for fiscal quarter when validating the quarter?
> > 
> > 
> > i'm not sure if fiscal quarter can be validated by itself. generally speaking, fiscal year + fiscal quarter selections need to map to the calendar periods in the files. If they dont map correctly, the file should be rejected and not parsed. i think this looks a bit different by section:
> > 
> > * section 1: FY2024Q1 (frontend selections) means `RPT_MONTH_YEAR` in the T1/M2, T2/M2, and M3/T3 records should have values equal to 202310, 202311 or 202312, and so on...
> > * section 2: similar to Section 1 but  for T4/M4 and T5/M5 records
> > * section 3: FY2024Q1 means `CALENDAR_QUARTER` in the T6/M6 record should have a value equal to 20234, and so on...
> > * section 4: similar to section 3 but for T7/M7 records
> 
> @ADPennington After discussing with the team, we decided to compare the HEADER year and quarter with frontend date, and draft another ticket for further work to compare each field with frontend date.

sounds good @raftmsohani; thank you.  please note the header year/quarter represent a calendar period, and frontend year/qtr represents a fiscal period.

File: tdrs-backend/tdpservice/parsers/fields.py [hunk 1]
```diff
@@ -1,10 +1,19 @@
...
+def value_is_empty(value, length):
+    """Handle 'empty' values as field inputs."""
+    empty_values = [
+        '',
+        ' '*length,  # '     '
+        '#'*length,  # '#####'
+        '_'*length,  # '_____'
+    ]
+
+    return value is None or value in empty_values
```
Comment (raftmsohani) [inline tdrs-backend/tdpservice/parsers/fields.py]: I had to add value_is_empty to a couple of files where I needed to use to resolve circular import. I am fixing this issue

File: tdrs-backend/tdpservice/parsers/schema_defs/ssp/m1.py [hunk 2]
```diff
@@ -1,9 +1,8 @@
 """Schema for SSP M1 record type."""
 
 
-from tdpservice.parsers.util import SchemaManager
```
Comment (raftmsohani) [inline tdrs-backend/tdpservice/parsers/schema_defs/ssp/m1.py:4]: resolving circular import

**Comment by raftmsohani** [comment]
As a follow up, have created this issue: https://github.com/raft-tech/TANF-app/issues/2799

**Review by andrew-jameson — APPROVED** [review]
LGTM

File: tdrs-backend/Dockerfile [hunk 3]
```diff
@@ -16,6 +16,7 @@ RUN apt-get -y update
 RUN apt-get -y upgrade
 # Install a new package:
 RUN apt-get install -y gcc && apt-get install -y graphviz && apt-get install -y graphviz-dev
+RUN apt-get install postgresql-client -y
```
Comment (elipe17) [inline tdrs-backend/Dockerfile:19]: Is this a convenience add?

**Review by elipe17 — APPROVED** [review]

File: tdrs-backend/Dockerfile (hunk 3 above)
Comment (raftmsohani) [reply tdrs-backend/Dockerfile:19]: yes, it is. Kinda unrelated to this PR but is nice to have

**Review by jtimpe — APPROVED** [review]

**Review by ADPennington — APPROVED** [review]
lgtm @raftmsohani 🚀 

Verified the following:

- pre-parsing error is raised when header cal period does not map properly to frontend fiscal periods
 e.g. `Submitted reporting year:2022, quarter:Q4 doesn't match file reporting year:2022, quarter:Q2.`
- this error is raised for all sections and all data types (TANF, SSP, Tribal TANF)
- file status is rejected. 

#2799  to deal with the scenario when other records have a reporting period that is not expected.

**Pull request merged**